public/friendlycode/js/fc/publisher.js

```diff
@@ -0,0 +1,203 @@
+    }
+
+    if(publisher.fsync) {
+      publisher.fsync.setCallback(publisher.handlers.unpublishedChangesPrompt);
+    }
+
+    interface.buttons.publish.on("click", publisher.handlers.publish);
+  };
+
+  Publisher.prototype.publish = function() {
+    var publisher = this;
+    var interface = publisher.interface;
+
+    function setState(done) {
+      var buttons = interface.buttons;
+      var handlers = publisher.handlers;
+      var toggle = done ? "on" : "off";
+
+      buttons.publish[toggle]("click", handlers.publish);
+      buttons.update[toggle]("click", handlers.publish);
+      buttons.publish.text(done ? TEXT_PUBLISH : TEXT_PUBLISHING);
+      buttons.update.text(done ? TEXT_UPDATE_PUBLISH : TEXT_PUBLISHING);
+    }
+
+    // SaveAll
+
+    setState(false);
+
+    var request = publisher.generateRequest("/publish");
+    request.done(function(project) {
+      if(request.status !== 200) {
+        console.error("[Bramble] Server was unable to publish project, responded with status ", request.status);
+        return;
+      }
+
+      publisher.update(project.link, true);
+    });
+    request.fail(function(jqXHR, status, err) {
+      console.error("[Bramble] Failed to send request to publish project to the server with: ", err);
+    });
+    request.always(function() {
+      setState(true);
+    });
+  };
+
+  Publisher.prototype.unpublish = function() {
+    var publisher = this;
+    var handlers = publisher.handlers;
+    var interface = publisher.interface;
+    var buttons = interface.buttons;
+
+    function setState(done) {
+      buttons.publish[done ? "on" : "off"]("click", handlers.publish);
+      buttons.unpublish.children("span").text(done ? TEXT_UNPUBLISH : TEXT_UNPUBLISHING);
+    }
+
+    // Disable all actions during the unpublish
+    buttons.unpublish.off("click", handlers.unpublish);
+    setState(false);
+
+    var request = publisher.generateRequest("/unpublish");
+    request.done(function() {
+      if(request.status !== 200) {
+        console.error("[Bramble] Server was unable to unpublish project, responded with status ", request.status);
+        buttons.unpublish.on("click", handlers.unpublish);
+        return;
+      }
+
+      publisher.update("");
+    });
+    request.fail(function(jqXHR, status, err) {
+      console.error("[Bramble] Failed to send request to unpublish project to the server with: ", err);
+      buttons.unpublish.on("click", handlers.unpublish);
+    });
+    request.always(function() {
+      setState(true);
+    });
+  };
+
+  Publisher.prototype.generateRequest = function(route) {
+    var publisher = this;
+
+    return $.ajax({
+      contentType: "application/json",
+      headers: {
+        "X-Csrf-Token": publisher.csrfToken,
+        "Accept": "application/json"
+      },
+      type: "PUT",
+      url: host + route,
+      data: JSON.stringify({
+        description: publisher.interface.description.val() || " ",
+        public: publisher.public,
```

Comment:
This might be clearer if the value was stored/toggled in the html element, and pulled from that element here.